### PR TITLE
AdditionalCommands support in kubeadm spec

### LINKED ISSUE #108: Add AdditionalCommands field to KubeadmConfigSpec
/kind feature

**Describe the solution you'd like**
`ControlPlaneInput` and `NodeInput` both support `AdditionalCommands` in their teamplate and struct but it looks like `KubeadmConfigSpec` does not have a field that the user can use to take advantage of `AdditionalCommands`.

The `AdditionalCommands` field in `KubeadmConfigSpec` should eventually get wired up to `ControlPlaneInput` and `NodeInput`.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

### PR DESCRIPTION
**Pull request by ykakarap**
**What this PR does / why we need it**:
This PR adds `AdditionalCommands` to Kubeadm spec and wires it up in the reconciler

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #108 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

/cc @chuckha

### CODE REVIEW

File: api/v1alpha2/kubeadmbootstrapconfig_types.go [hunk 1]
```diff
@@ -36,6 +36,9 @@ type KubeadmConfigSpec struct {
 	// AdditionalUserDataFiles specifies extra files to be passed to user_data upon creation.
 	// +optional
 	AdditionalUserDataFiles []Files `json:"additionalUserDataFiles,omitempty"`
+	// AdditionalCommands specifies extra commands to be passed to user_data upon creation
+	// +optional
+	AdditionalCommands []string `json:"additionalCommands,omitempty"`
```
Comment (detiber) [inline api/v1alpha2/kubeadmbootstrapconfig_types.go]: I know previously we had discussed having both pre and post commands rather than just a single AdditionalCommands field. Could you update this for that? Maybe `AdditionalPreCommands` and `AdditionalPostCommands`?

Updating this file will also require updating the generated CRDs.

Comment (ykakarap) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: Supporting `AdditionalPreCommands`  and `AdditionalPostCommands` will also need changes to the cloud-init template, correct?

Something like this:
```
runcmd:
{{- template "commands" .AdditionalPreCommands }}
  - 'kubeadm init --config /tmp/kubeadm.yaml'
{{- template "commands" .AdditionalPostCommands }}
```

Comment (detiber) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: That would be correct, yes.

Comment (ykakarap) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: @detiber I was wondering if there an automatic way to generate the updated YAML using the updated type or maybe the reverse, generate the type using the YAML?

Comment (chuckha) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: if you modify the struct and run make generate the yaml should all get updated (not the cloud-init stuff though, that will be manual)

**Comment by ykakarap** [comment]
@chuckha @detiber PR is ready for review

File: api/v1alpha2/kubeadmbootstrapconfig_types.go [hunk 2]
```diff
@@ -36,6 +36,14 @@ type KubeadmConfigSpec struct {
 	// AdditionalUserDataFiles specifies extra files to be passed to user_data upon creation.
 	// +optional
 	AdditionalUserDataFiles []Files `json:"additionalUserDataFiles,omitempty"`
+	// AdditionalPreCommands specifies extra commands to be passed to user_data upon creation
+	// which run before kubeadm runs
+	// +optional
+	AdditionalPreCommands []string `json:"additionalPreCommands,omitempty"`
```
Comment (vincepri) [inline api/v1alpha2/kubeadmbootstrapconfig_types.go]: What do we think about dropping the `Additional` prefix for these fields? Given that we have `Pre` and `Post` seems that should be fine?

File: cloudinit/cloudinit.go [hunk 3]
```diff
@@ -18,9 +18,10 @@ package cloudinit
 
 import (
 	"bytes"
-	"sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
 	"text/template"
 
+	"sigs.k8s.io/cluster-api-bootstrap-provider-kubeadm/api/v1alpha2"
+
```
Comment (vincepri) [inline cloudinit/cloudinit.go]: Nit: Extra space

File: cloudinit/cloudinit.go [hunk 4]
```diff
@@ -32,10 +33,11 @@ const (
 
 // BaseUserData is shared across all the various types of files written to disk.
 type BaseUserData struct {
-	Header             string
-	AdditionalCommands []string
-	AdditionalFiles    []v1alpha2.Files
-	WriteFiles         []v1alpha2.Files
+	Header                 string
+	AdditionalPreCommands  []string
+	AdditionalPostCommands []string
```
Comment (vincepri) [inline cloudinit/cloudinit.go]: Same here re:Additional prefix

File: api/v1alpha2/kubeadmbootstrapconfig_types.go [hunk 5]
```diff
@@ -36,6 +36,14 @@ type KubeadmConfigSpec struct {
 	// AdditionalUserDataFiles specifies extra files to be passed to user_data upon creation.
 	// +optional
 	AdditionalUserDataFiles []Files `json:"additionalUserDataFiles,omitempty"`
+	// AdditionalPreCommands specifies extra commands to be passed to user_data upon creation
```
Comment (detiber) [inline api/v1alpha2/kubeadmbootstrapconfig_types.go]: Not directly related to this PR, but we should likely avoid references to user_data, since that is an AWS specific term. @chuckha thoughts?

File: api/v1alpha2/kubeadmbootstrapconfig_types.go (hunk 2 above)
Comment (chuckha) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: Interesting, i wonder why we had the word Additional commands in the first place? I guess to signify in addition to kubeadm init/join.

Does Additional add anything to this field? I think Additional adds a sense that these fields are optional. That if I don't need any additional commands I don't have to have them. But if they were just called Pre and Post I might feel more inclined like I need them.

Overall, I don't have any strong feelings and am happy with both Additional and Pre/Post. Up to @ykakarap if he'd like to update it or not.

File: api/v1alpha2/kubeadmbootstrapconfig_types.go (hunk 5 above)
Comment (chuckha) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: Agreed, it should be `AdditionalFiles` I think, or perhaps simply `Files`. But yes, out of scope for this PR.

File: api/v1alpha2/kubeadmbootstrapconfig_types.go (hunk 2 above)
Comment (ykakarap) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: Additional signifies that it runs in addition to kubeadm init/join command as mentioned by @chuckha .

If we want to drop "Additional" I suggest naming them `PreKubeadmCommands` and `PostKubeadmCommands`. This will suggest that they are commands besides the kubeadm commands and are also more descriptive of their behavior.

Thoughts?

Comment (chuckha) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: That also works. I was almost worried about tying the name to an implementation and then remembered this is the kubeadm bootstrap provider :)

I have no opinion on this. I'm happy to merge with Additional or Pre/Post KubeadmCommands

Comment (chuckha) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: @vincepri do you have strong feelings here?

Comment (fabriziopandini) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: I have not strong feeling, but If I have to express a preference is
- AdditionalFiles --> Files
- AdditionalPreCommands --> PreKubeadmCommands
- AdditionalPostCommands --> PostKubeadmCommands

Comment (vincepri) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: @fabriziopandini's suggestion looks good to me :)

File: api/v1alpha2/kubeadmbootstrapconfig_types.go (hunk 5 above)
Comment (vincepri) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: I'd rephrase with something like 
```golang
... specifies extra commands to be run before kubeadm.
```

File: api/v1alpha2/kubeadmbootstrapconfig_types.go (hunk 2 above)
Comment (ykakarap) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: >AdditionalFiles --> Files

`AdditionalFiles` is in `BaseUserData` struct. Shall i only change that to `Files` or also change the `AdditionalUserDataFiles` to `UserDataFiles` in `KubeadmConfigSpec` ?

Comment (chuckha) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: The important change is the KubeadmConfigSpec.

Ideally we don't reference UserData anywhere as that is a legacy AWS thing, but we can do that in a different PR and not change the API.

Comment (chuckha) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: To be explicit, please change `AdditionalUserDataFiles` to `Files` on the `KubeadmConfigSpec` type

Comment (ykakarap) [reply api/v1alpha2/kubeadmbootstrapconfig_types.go]: Renamed the fields as discussed :)

**Comment by ykakarap** [comment]
@chuckha @vincepri @fabriziopandini This PR is ready for review.
